admission_control/admission-control-service/src/upstream_proxy.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::PeerId;
+use config::config::{AdmissionControlConfig, NodeConfig};
+use failure::prelude::*;
+use futures::{
+    channel::mpsc,
+    stream::{select_all, StreamExt},
+};
+use logger::prelude::*;
+use network::{
+    proto::{AdmissionControlMsg, SubmitTransactionRequest, SubmitTransactionResponse},
+    validator_network::{AdmissionControlNetworkEvents, AdmissionControlNetworkSender, Event},
+};
+use rand::seq::SliceRandom;
+use std::{collections::HashMap, str::FromStr};
+
+#[derive(Clone)]
+struct PeerSyncState {
+    is_alive: bool,
+}
+
+type PeerInfo = HashMap<PeerId, PeerSyncState>;
+
+/// Struct that owns all dependencies required by upstream proxy routine
+pub(crate) struct UpstreamProxy {
+    ac_config: AdmissionControlConfig,
+    network_sender: AdmissionControlNetworkSender,
+    peer_info: PeerInfo,
+    // used to process client requests
+    client_events: mpsc::UnboundedReceiver<AdmissionControlMsg>,
+    is_validator: bool,
+}
+
+impl UpstreamProxy {
+    /// bootstrap of UpstreamProxy
+    pub fn new(
+        config: &NodeConfig,
+        network_sender: AdmissionControlNetworkSender,
+        client_events: mpsc::UnboundedReceiver<AdmissionControlMsg>,
+        is_validator: bool,
+    ) -> Self {
+        let upstream_peers: HashMap<_, _> = config
+            .state_sync
+            .upstream_peers
+            .upstream_peers
+            .iter()
+            .map(|peer_id_str| {
+                (
+                    PeerId::from_str(peer_id_str).unwrap_or_else(|_| {
+                        panic!("Failed to parse peer_id from string: {}", peer_id_str)
+                    }),
+                    PeerSyncState { is_alive: true },
+                )
+            })
+            .collect();
+        Self {
+            ac_config: config.admission_control.clone(),
+            network_sender,
+            peer_info: upstream_peers,
+            client_events,
+            is_validator,
+        }
+    }
+
+    /// main routine. starts sync coordinator that listens for CoordinatorMsg
+    pub async fn process_network_messages(
+        mut self,
+        network_events: Vec<AdmissionControlNetworkEvents>,
+    ) {
+        let mut events = select_all(network_events).fuse();
+
+        loop {
+            ::futures::select! {
+                mut msg = self.client_events.select_next_some() => {
+                    if msg.has_submit_transaction_request() {
```

Comment:
this looks unnecessary. Can client_events use `SubmitTransactionRequest` from ac proto definition instead?